rate-limiter/src/rate_limiter.rs

```diff
@@ -53,9 +36,23 @@ impl SleepingRateLimiter {
             "Rate-Limiter attempting to read {}.",
             read_size
         );
-        if let Some(sleep) = self.set_sleep(read_size) {
-            sleep.await;
-        }
+
+        let now = Instant::now();
+        let next_wait = self.rate_limiter.rate_limit(read_size, now);
+        let wait_until = if let Some(next_wait) = next_wait {
+            now + next_wait
+        } else {
+            return self;
+        };
```

Comment:
This code looks a bit convoluted imo, maybe just
```
if let Some(wait_until) = [rate_limit].map(|next_wait| now+next_wait) {
 //trace+sleep
}
self
```

also, why `sleep_until` rather than just sleep for however much `rate_limit` returns?

Reply:
That was artifact from previous version and refactoring. Thanks for noting.